File: internal/broker/fence_watcher.go
package broker

import (
	"context"
	"encoding/json"
	"errors"
	"io/fs"
	"log/slog"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/woestebanaan/skafka/internal/protocol/handlers"
)

// FenceWatcher polls the producer-fences directory on the shared
// RWX PVC and applies any new fence entries to the local storage
// engine via ProducerEpochFencer.FenceProducerEpoch. Closes the
// cross-broker fence gap (gh #108 phase 2): when a coordinator
// broker bumps a producer's epoch on InitProducerId, partitions
// led by *other* brokers see no signal until the new session
// writes there — an in-flight zombie from the old session can
// slip through. The watcher applies every peer's outbound fences
// so every broker's storage view of every PID's epoch monotonically
// catches up to the cluster max.
//
// Read-only: the watcher never writes. Self-loop avoidance is
// enforced by skipping the file whose name matches our own
// brokerID — fences this broker emitted are already applied
// locally by the in-process call to FenceProducerEpoch.
//
// Idempotent: per-file (PID → highest-epoch-applied) cache means
// a fence already applied isn't re-applied on the next tick.
// Idempotency at the engine layer is also guarded
// (FenceProducerEpoch is a no-op when entry.epoch >= epoch),
// but tracking here saves the engine-wide RLock loop on every
// tick when nothing changed.
type FenceWatcher struct {
	dir       string
	selfFile  string // basename of our own outbound fence file (skip)
	fencer    handlers.ProducerEpochFencer
	pollEvery time.Duration

	mu      sync.Mutex
	applied map[string]map[int64]int16 // peerFile → PID → highest epoch applied
}

// NewFenceWatcher constructs a watcher rooted at the producer-
// fences directory. selfFile is the basename ("from-<brokerID>.json")
// the watcher will skip — its content is fences this broker
// emitted, already applied in-process.
func NewFenceWatcher(dir, selfFile string, fencer handlers.ProducerEpochFencer) *FenceWatcher {
	return &FenceWatcher{
		dir:       dir,
		selfFile:  selfFile,
		fencer:    fencer,
		pollEvery: 2 * time.Second,
		applied:   make(map[string]map[int64]int16),
	}
}

// Run drives the polling loop until ctx is cancelled. Errors
// reading individual peer files are logged and the loop continues
// — a half-written file under a peer's tmp+rename is transient
// and the next tick picks it up.
func (w *FenceWatcher) Run(ctx context.Context) {
	t := time.NewTicker(w.pollEvery)
	defer t.Stop()
	// Tick once on entry so a peer's pre-existing file (e.g. after
	// our own restart) is applied without waiting a full interval.
	w.Tick()
	for {
		select {
		case <-ctx.Done():
			return
		case <-t.C:
			w.Tick()
		}
	}
}

// Tick is a single-pass scan of the fence directory. Exported so
// integration tests can drive the watcher synchronously instead
// of waiting for the 2s poll interval. Run() also calls this on
// entry and on every ticker fire.
func (w *FenceWatcher) Tick() {
	entries, err := os.ReadDir(w.dir)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return // dir doesn't exist yet — no fences emitted anywhere
		}
		slog.Warn("fence watcher: read dir", "dir", w.dir, "err", err)
		return
	}
	for _, e := range entries {
		name := e.Name()
		if e.IsDir() {
			continue
		}
		if !strings.HasPrefix(name, "from-") || !strings.HasSuffix(name, ".json") {
			continue
		}
		if name == w.selfFile {
			continue
		}
		w.applyPeer(name)
	}
}

func (w *FenceWatcher) applyPeer(name string) {
	path := filepath.Join(w.dir, name)
	data, err := os.ReadFile(path)
	if err != nil {
		if !errors.Is(err, fs.ErrNotExist) {
			slog.Warn("fence watcher: read peer", "file", path, "err", err)
		}
		return
	}
	if len(data) == 0 {
		return
	}
	var state map[string]int16
	if err := json.Unmarshal(data, &state); err != nil {
		slog.Warn("fence watcher: decode peer", "file", path, "err", err)
		return
	}

	w.mu.Lock()
	last, ok := w.applied[name]
	if !ok {
		last = make(map[int64]int16)
		w.applied[name] = last
	}
	pending := make(map[int64]int16, len(state))
	for k, epoch := range state {
		pid, err := strconv.ParseInt(k, 10, 64)
		if err != nil {
			continue
		}
		if existing, ok := last[pid]; ok && existing >= epoch {
			continue
		}
		pending[pid] = epoch
		last[pid] = epoch
	}
	w.mu.Unlock()

	for pid, epoch := range pending {
		w.fencer.FenceProducerEpoch(pid, epoch)
	}
}

File: crates/kaas-broker/src/fence_watcher.rs
//! Cross-broker producer-epoch fence watcher (gh #108 phase 2).
//!
//! Polls `<data_dir>/__cluster/producer_fences/` every 2 s, reads
//! every peer broker's `from-<id>.json`, and for each
//! `(pid, epoch)` pair that's newer than what we've already
//! applied calls [`ProducerEpochFencer::fence_producer_epoch`].
//! Self-loop avoidance: the file named `from-<our broker id>.json`
//! is skipped — we already applied those fences in-process when
//! `InitProducerId` ran.
//!
//! **Why poll, not `notify`?** The Phase 5 acknowledgement applies
//! here too: Linux `inotify` does not fire for changes made by
//! other NFS clients, so a `notify`-driven watcher would silently
//! miss every peer broker's writes on a shared-RWX volume. The
//! 2 s mtime poll is the load-bearing mechanism; see gh #166 for
//! the doc gap.

use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::Arc;
use std::time::Duration;

use parking_lot::Mutex;
use tokio::time::interval;
use tokio_util::sync::CancellationToken;

/// Hook the watcher calls to apply a `(pid, epoch)` fence into the
/// local storage view. Production wires this to the storage engine's
/// per-partition `producer_states::fence` walker so a zombie batch
/// from the old session is rejected even on partitions the new
/// session hasn't yet touched. Tests substitute a capturing impl.
///
/// Same shape as the in-process callback `InitProducerIdHandler`
/// uses for the local fence (workstream C — currently a TODO until
/// the engine grows a cross-partition `fence_producer_epoch`).
pub trait ProducerEpochFencer: Send + Sync + 'static {
    fn fence_producer_epoch(&self, pid: i64, epoch: i16);
}

/// Default poll interval: 2 s.
pub const DEFAULT_POLL: Duration = Duration::from_secs(2);

/// Reads every peer broker's outbound fence file and dispatches new
/// entries to the wired [`ProducerEpochFencer`]. Designed to be
/// `tokio::spawn`ed under a cluster-runtime cancellation token.
pub struct FenceWatcher {
    dir: PathBuf,
    self_file: String,
    fencer: Arc<dyn ProducerEpochFencer>,
    poll: Duration,
    /// Per-peer `(pid → highest epoch already applied)` cache so
    /// the engine isn't re-fenced for the same value every tick.
    applied: Mutex<HashMap<String, HashMap<i64, i16>>>,
}

impl std::fmt::Debug for FenceWatcher {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("FenceWatcher")
            .field("dir", &self.dir)
            .field("self_file", &self.self_file)
            .field("poll", &self.poll)
            .finish()
    }
}

impl FenceWatcher {
    /// `dir` is the conventional `producer_fences/` directory;
    /// `self_broker_id` derives the file name to skip
    /// (`from-<self_broker_id>.json`).
    pub fn new(dir: PathBuf, self_broker_id: &str, fencer: Arc<dyn ProducerEpochFencer>) -> Self {
        Self {
            dir,
            self_file: format!("from-{self_broker_id}.json"),
            fencer,
            poll: DEFAULT_POLL,
            applied: Mutex::new(HashMap::new()),
        }
    }

    /// Override the poll interval. Test-only — production keeps the
    /// 2 s default.
    pub fn with_poll(mut self, d: Duration) -> Self {
        self.poll = d;
        self
    }

    /// Drive the polling loop until `cancel` fires. Ticks once on
    /// entry so a peer's pre-existing file (e.g. after our own
    /// restart) is applied without waiting a full interval.
    pub async fn run(self: Arc<Self>, cancel: CancellationToken) {
        self.tick();
        let mut t = interval(self.poll);
        // Skip the immediate tick `interval` always yields first.
        t.tick().await;
        loop {
            tokio::select! {
                () = cancel.cancelled() => return,
                _ = t.tick() => self.tick(),
            }
        }
    }

    /// Single synchronous pass. Exposed so integration tests can
    /// drive the watcher deterministically.
    pub fn tick(&self) {
        let entries = match fs::read_dir(&self.dir) {
            Ok(e) => e,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return,
            Err(e) => {
                tracing::warn!(
                    dir = %self.dir.display(),
                    %e,
                    "fence watcher: scanning the cross-broker fence directory failed; \
                     peer producer-epoch bumps will not propagate this cycle",
                );
                return;
            }
        };
        for entry in entries.flatten() {
            let name = entry.file_name();
            let name = match name.to_str() {
                Some(n) => n,
                None => continue,
            };
            if !name.starts_with("from-") || !name.ends_with(".json") {
                continue;
            }
            if name == self.self_file {
                continue;
            }
            self.apply_peer(name);
        }
    }

    fn apply_peer(&self, name: &str) {
        let path = self.dir.join(name);
        let data = match fs::read(&path) {
            Ok(d) => d,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return,
            Err(e) => {
                tracing::warn!(
                    file = %path.display(),
                    %e,
                    "fence watcher: reading peer broker's fence file failed; \
                     this broker may miss a producer-epoch bump until the next tick",
                );
                return;
            }
        };
        if data.is_empty() {
            return;
        }
        let peer_state: HashMap<String, i16> = match serde_json::from_slice(&data) {
            Ok(s) => s,
            Err(e) => {
                tracing::warn!(
                    file = %path.display(),
                    %e,
                    "fence watcher: decoding peer broker's fence-state JSON failed; \
                     file likely mid-write or corrupted; will retry next tick",
                );
                return;
            }
        };

        let mut pending: Vec<(i64, i16)> = Vec::new();
        {
            let mut applied = self.applied.lock();
            let last = applied.entry(name.to_owned()).or_default();
            for (k, &epoch) in &peer_state {
                let pid: i64 = match k.parse() {
                    Ok(p) => p,
                    Err(_) => continue,
                };
                if let Some(&existing) = last.get(&pid) {
                    if existing >= epoch {
                        continue;
                    }
                }
                pending.push((pid, epoch));
                last.insert(pid, epoch);
            }
        }
        for (pid, epoch) in pending {
            self.fencer.fence_producer_epoch(pid, epoch);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[derive(Default)]
    struct Capturing {
        calls: parking_lot::Mutex<Vec<(i64, i16)>>,
    }

    impl ProducerEpochFencer for Capturing {
        fn fence_producer_epoch(&self, pid: i64, epoch: i16) {
            self.calls.lock().push((pid, epoch));
        }
    }

    fn write_peer(dir: &Path, name: &str, state: &HashMap<i64, i16>) {
        // Stringified keys to match the on-disk shape.
        let stringy: HashMap<String, i16> =
            state.iter().map(|(k, v)| (k.to_string(), *v)).collect();
        let data = serde_json::to_vec(&stringy).unwrap();
        fs::write(dir.join(name), data).unwrap();
    }

    #[test]
    fn applies_peer_fences() {
        let tmp = tempfile::tempdir().unwrap();
        let capturing = Arc::new(Capturing::default());
        let w = FenceWatcher::new(tmp.path().to_path_buf(), "kaas-0", capturing.clone());
        let mut peer = HashMap::new();
        peer.insert(42, 3);
        peer.insert(99, 1);
        write_peer(tmp.path(), "from-kaas-1.json", &peer);
        w.tick();
        let mut calls = capturing.calls.lock().clone();
        calls.sort();
        assert_eq!(calls, vec![(42, 3), (99, 1)]);
    }

    #[test]
    fn skips_self_file() {
        let tmp = tempfile::tempdir().unwrap();
        let capturing = Arc::new(Capturing::default());
        let w = FenceWatcher::new(tmp.path().to_path_buf(), "kaas-0", capturing.clone());
        let mut state = HashMap::new();
        state.insert(42, 3);
        write_peer(tmp.path(), "from-kaas-0.json", &state); // self
        write_peer(tmp.path(), "from-kaas-1.json", &state); // peer
        w.tick();
        // Only the peer entry fires.
        assert_eq!(capturing.calls.lock().len(), 1);
    }

    #[test]
    fn dedupe_across_ticks() {
        let tmp = tempfile::tempdir().unwrap();
        let capturing = Arc::new(Capturing::default());
        let w = FenceWatcher::new(tmp.path().to_path_buf(), "kaas-0", capturing.clone());
        let mut state = HashMap::new();
        state.insert(42, 3);
        write_peer(tmp.path(), "from-kaas-1.json", &state);
        w.tick();
        w.tick();
        assert_eq!(capturing.calls.lock().len(), 1);
    }

    #[test]
    fn higher_epoch_after_first_apply_fires_once_more() {
        let tmp = tempfile::tempdir().unwrap();
        let capturing = Arc::new(Capturing::default());
        let w = FenceWatcher::new(tmp.path().to_path_buf(), "kaas-0", capturing.clone());
        let mut state = HashMap::new();
        state.insert(42, 3);
        write_peer(tmp.path(), "from-kaas-1.json", &state);
        w.tick();
        state.insert(42, 5); // bump
        write_peer(tmp.path(), "from-kaas-1.json", &state);
        w.tick();
        let calls = capturing.calls.lock().clone();
        assert_eq!(calls, vec![(42, 3), (42, 5)]);
    }

    #[test]
    fn missing_dir_is_silent_noop() {
        let tmp = tempfile::tempdir().unwrap();
        let missing = tmp.path().join("does-not-exist");
        let capturing = Arc::new(Capturing::default());
        let w = FenceWatcher::new(missing, "kaas-0", capturing.clone());
        w.tick(); // must not panic / error
        assert!(capturing.calls.lock().is_empty());
    }

    #[test]
    fn corrupt_peer_file_is_skipped() {
        let tmp = tempfile::tempdir().unwrap();
        fs::write(tmp.path().join("from-kaas-1.json"), b"\xffnot-json").unwrap();
        let capturing = Arc::new(Capturing::default());
        let w = FenceWatcher::new(tmp.path().to_path_buf(), "kaas-0", capturing.clone());
        w.tick();
        assert!(capturing.calls.lock().is_empty());
    }

    #[test]
    fn ignores_files_not_matching_pattern() {
        let tmp = tempfile::tempdir().unwrap();
        // Wrong prefix / suffix.
        let mut state = HashMap::new();
        state.insert(42, 3);
        write_peer(tmp.path(), "from-kaas-1.json", &state);
        let stringy: HashMap<String, i16> =
            state.iter().map(|(k, v)| (k.to_string(), *v)).collect();
        fs::write(
            tmp.path().join("notes.txt"),
            serde_json::to_vec(&stringy).unwrap(),
        )
        .unwrap();
        let capturing = Arc::new(Capturing::default());
        let w = FenceWatcher::new(tmp.path().to_path_buf(), "kaas-0", capturing.clone());
        w.tick();
        assert_eq!(capturing.calls.lock().len(), 1);
    }
}
